[@property] Initial value cleanups

anttijk opened the pull request:
#### 0e9c19d6b9a6b5d1604dbe2f531b147eb0a5c9f9
<pre>
[@property] Initial value cleanups
<a href="https://bugs.webkit.org/show_bug.cgi?id=250707">https://bugs.webkit.org/show_bug.cgi?id=250707</a>
rdar://104330630

Reviewed by Antoine Quint.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendCustomProperty):
(WebCore::CSSPropertyAnimation::propertyRequiresBlendingForAccumulativeIteration):
(WebCore::CSSPropertyAnimation::propertiesEqual):
(WebCore::CSSPropertyAnimation::canPropertyBeInterpolated):
* Source/WebCore/css/CSSRegisteredCustomProperty.cpp:
(WebCore::CSSRegisteredCustomProperty::CSSRegisteredCustomProperty): Deleted.
(WebCore::CSSRegisteredCustomProperty::initialValueCopy const): Deleted.
* Source/WebCore/css/CSSRegisteredCustomProperty.h:
(WebCore::CSSRegisteredCustomProperty::initialValue const): Deleted.

Just make the field public.

* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue):

Remove the dead leftover code for resolving Length values.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::changedCustomPaintWatchedProperty):
(WebCore::RenderStyle::customPropertyValue const):

Handle resolving the initial value for a registered property here. Clients have to provide the registry.

(WebCore::RenderStyle::customPropertyValueWithoutResolvingInitial const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::getCustomProperty const): Deleted.
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::get const):
(WebCore::Style::CustomPropertyRegistry::isInherited const):
* Source/WebCore/style/CustomPropertyRegistry.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialCustomProperty):

Canonical link: <a href="https://commits.webkit.org/258991@main">https://commits.webkit.org/258991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60adfbe82969d892ecb502daac91f108d3abea68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103560 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112796 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173002 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107513 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3575 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95817 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111968 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10554 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38290 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92375 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25214 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79932 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6055 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26616 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6235 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3146 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12217 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46126 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6177 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7988 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -4025,40 +4025,30 @@ static Ref<CSSCustomPropertyValue> blendedCSSCustomPropertyValue(const RenderSty
-        }
-        return nullptr;
-    };
+    // FIXME: When would there be no document?
```
You could add an `ASSERT()` and find out from tests :)

anttijk replied:
I was more of hoping that document would be a non-nullable type.

graouts commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2670,6 +2669,25 @@ void RenderStyle::setNonInheritedCustomPropertyValue(const AtomString& name, Ref
     m_rareNonInheritedData.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
 }
 
+const CSSCustomPropertyValue* RenderStyle::customPropertyValue(const AtomString& name, const Style::CustomPropertyRegistry& registry) const
+{
+    if (auto* value = customPropertyValueWithoutResolvingInitial(name))
+        return value;
+
+    // Alternatively we could just initialize the registered initial values to each RenderStyle.
+    auto* registered = registry.get(name);
+    return registered ? registered->initialValue.get() : nullptr;
+}
+
+const CSSCustomPropertyValue* RenderStyle::customPropertyValueWithoutResolvingInitial(const AtomString& name) const
+{
+    for (auto* map : { &nonInheritedCustomProperties(), &inheritedCustomProperties() }) {
+        if (auto* value = map->get(name))
+            return value;
+    }
+    return nullptr;
+}
```
Maybe this could be a single method where the `CustomPropertyRegistry` is a pointer and you only check for an initial value if the registry is provided?

anttijk replied:
I had that initially but I don't want people wondering what happens if they don't provide a registry (and they always should).

graouts reviewed (approved):
I hope this is an intermediary step towards not needing the registry at all for initial values and for those to be baked into `RenderStyle`.

anttijk committed:
[@property] Initial value cleanups
https://bugs.webkit.org/show_bug.cgi?id=250707
rdar://104330630

Reviewed by Antoine Quint.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendCustomProperty):
(WebCore::CSSPropertyAnimation::propertyRequiresBlendingForAccumulativeIteration):
(WebCore::CSSPropertyAnimation::propertiesEqual):
(WebCore::CSSPropertyAnimation::canPropertyBeInterpolated):
* Source/WebCore/css/CSSRegisteredCustomProperty.cpp:
(WebCore::CSSRegisteredCustomProperty::CSSRegisteredCustomProperty): Deleted.
(WebCore::CSSRegisteredCustomProperty::initialValueCopy const): Deleted.
* Source/WebCore/css/CSSRegisteredCustomProperty.h:
(WebCore::CSSRegisteredCustomProperty::initialValue const): Deleted.

Just make the field public.

* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue):

Remove the dead leftover code for resolving Length values.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::changedCustomPaintWatchedProperty):
(WebCore::RenderStyle::customPropertyValue const):

Handle resolving the initial value for a registered property here. Clients have to provide the registry.

(WebCore::RenderStyle::customPropertyValueWithoutResolvingInitial const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::getCustomProperty const): Deleted.
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::get const):
(WebCore::Style::CustomPropertyRegistry::isInherited const):
* Source/WebCore/style/CustomPropertyRegistry.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialCustomProperty):

Canonical link: https://commits.webkit.org/258991@main
